### ISSUE TITLE
Start looks to be failing to close stdin/stdout properly

### ISSUE THREAD

**Issue by clintonskitson** [emccode/rexray#85]
We just ran into a situation where when starting rexray with `sudo rexray start` remotely through SSH, `docker machine ..blah.. sudo rexray start` it would hang indefinitely.

In troubleshooting I tried running something similar locally, `sudo rexray start | sudo tee /tmp/temp`.  The tee is interpreting the stdin/stdout's and waiting for them to close apparently so acting similar to the remote SSH.  If you run that command it hangs.

**Comment by akutz** [comment]
Hi @clintonskitson / @kacole2,

Ah, the callback. So just tell SSH to use a pseudo-terminal:

```
[0]akutz@pax:rexray$ ssh -t 192.168.120.16 sudo rexray start --force
Starting REX-Ray...SUCESS!

  The REX-Ray daemon is now running at PID 23001. To
  shutdown the daemon execute the following command:

    sudo /usr/bin/rexray stop

Connection to 192.168.120.16 closed.
```

@kacole2, ignore the `--force` flag. It's a new feature of the `unstable` release that allows you to start the service whether or not any drivers are properly configured. It's primarily for testing.

However, the above will still fail on non SystemD systems. On SystemD systems you should be doing `ssh HOST -c sudo systemctl start rexray` anyway. @clintonskitson, please file a new issue requesting that the `start` and `stop` commands of the binary defer to `SystemD` and `SystemV` when present.

On a `SystemV` system there's an issue. The command I showed you above does start the service, but it stops as soon as the SSH session closes. This is because the service is running in *your* session. You see, Go binaries [cannot be forked](https://groups.google.com/forum/#!topic/golang-nuts/beKfn7rujNg). 

That means they're attached to the process from which they were spawned as child processes and will be killed when the parent process exits. Like an SSH session.

SystemD combats this easily enough, but Go services on SystemV are an issue if you want to remotely activate them with SSH. There is an [entire reddit thread discussing the issue](https://www.reddit.com/r/golang/comments/35v5bm/best_way_to_run_go_server_as_a_daemon/). One answer, use SystemD.

I've been playing around with `screen` to see if I can get it working, and I can, but not in a scriptable fashion.

```
[0]akutz@pax:rexray$ ssh -t 192.168.120.16 screen sudo rexray start -f --force
```

Then use `Ctrl-A d` to detach from the session. The issue is that `sudo` requires a TTY, but by allocating one you're telling `screen` and a `go` program to behave a certain way.

This is just a really crappy situation.

**Comment by akutz** [comment]
Hi @clintonskitson / @kacole2,

From [this link](http://stackoverflow.com/questions/161252/bash-start-multiple-chained-commands-in-background) I put this together that works:

```
[0]akutz@pax:rexray$ ssh 192.168.120.16 rexray status
REX-Ray is stopped

[0]akutz@pax:rexray$ `(sleep 2; ssh 192.168.120.16 screen -d rexray) > /dev/null &`; ssh -t 192.168.120.16 screen -S rexray sudo rexray start -f --force
[screen is terminating]
Connection to 192.168.120.16 closed.

[0]akutz@pax:rexray$ ssh 192.168.120.16 rexray status
REX-Ray is running at pid 31816

[0]akutz@pax:rexray$ ssh -t 192.168.120.16 sudo rexray stop
Shutting down REX-Ray...SUCCESS!
Connection to 192.168.120.16 closed.

[0]akutz@pax:rexray$ ssh 192.168.120.16 rexray status
REX-Ray is stopped
```

**Comment by clintonskitson** [comment]
@akutz so it sounds like we have two situations here.. 1) getting rexray to run and not exit with ssh 2) the fact that a start hangs due to stdin/stdout problem.

The #1 can be solved by having @kenny use the service command I assume instead of invoking the cli directly.  And #2, I'm not sure if it was addressed.

**Comment by akutz** [comment]
Hi @clintonskitson,

Neither issue as described is quite accurate. If I may...

## SSH Start
 `REX-Ray` runs fine and *stays* running when you start it with SSH as long as you're not trying to start it and exit immediately per the provided example. For that you need to use SystemD or some other system controller. All other cases fall under Go's inability to be forked. 

Here is me SSHd into an Ubuntu 14.04 where I start `REX-Ray`, log out of the host, SSH back into the host, and the service is still running:

```
[0]akutz@ubuntu:~$ sudo rexray status
REX-Ray is stopped

[0]akutz@ubuntu:~$ sudo rexray start --force
Starting REX-Ray...SUCESS!

  The REX-Ray daemon is now running at PID 32493. To
  shutdown the daemon execute the following command:

    sudo /usr/bin/rexray stop

[0]akutz@ubuntu:~$ tail /var/log/rexray/rexray.log 
 MMMM   "W" """"YUMMM,m"       "Mm, MMMM   "W" YMM   ""'mM"        

time="2015-09-24T13:08:59-05:00" level=info msg="dialing /tmp/ZpYcHRNGgtWPCwXLkpQlidYqMHXUyAIs.sock" 
time="2015-09-24T13:08:59-05:00" level=info msg="created pid file, pid=32493" 
time="2015-09-24T13:08:59-05:00" level=info msg="initialized module instance" address="tcp://:7979" id=1 ignoreFailOnInit=false typeId=1 typeName=AdminModule 
time="2015-09-24T13:08:59-05:00" level=warning msg="ignoring initialization failure" error.message="no storage drivers initialized" 
time="2015-09-24T13:08:59-05:00" level=warning msg="ignoring initialization failure" error.message="no storage drivers initialized" 
time="2015-09-24T13:08:59-05:00" level=info msg="started module" address="tcp://:7979" id=1 typeId=1 typeName=AdminModule 
time="2015-09-24T13:08:59-05:00" level=info msg="service sent registered modules start signals" 
time="2015-09-24T13:08:59-05:00" level=info msg="service successfully initialized, waiting on stop signal" 
[0]akutz@ubuntu:~$ exit
logout
Connection to 192.168.120.16 closed.
[0]akutz@pax:rexray$ ssh 192.168.120.16
Welcome to Ubuntu 14.04.3 LTS (GNU/Linux 3.19.0-25-generic x86_64)

 * Documentation:  https://help.ubuntu.com/

  System information as of Thu Sep 24 13:08:11 CDT 2015

  System load:  0.0               Processes:           111
  Usage of /:   4.4% of 61.65GB   Users logged in:     1
  Memory usage: 41%               IP address for eth0: 192.168.120.16
  Swap usage:   0%

  Graph this data and manage this system at:
    https://landscape.canonical.com/

9 packages can be updated.
7 updates are security updates.

Last login: Thu Sep 24 13:08:11 2015 from 192.168.120.3
[0]akutz@ubuntu:~$ sudo rexray status
REX-Ray is running at pid 32493
[0]akutz@ubuntu:~$ 
```

It's this that fails, but even in failure still *starts* the service:

```
[0]akutz@pax:rexray$ ssh -t 192.168.120.16 sudo "rexray start --force && rexray status"
Starting REX-Ray...SUCESS!

  The REX-Ray daemon is now running at PID 726. To
  shutdown the daemon execute the following command:

    sudo /usr/bin/rexray stop

REX-Ray is running at pid 726
Connection to 192.168.120.16 closed.
```

The fact is, there's probably some simple trick of which I'm unware that would make this work. I'm sure someone else could practice some Google Fu and figure it out...

## Stdout/Stdin
Piping `REX-Ray`'s stdout descriptor (I said piping, not redirecting) is not supported and thus this is breaking the program on purpose. It can (possibly) be addressed, but it's not high priority.

**Comment by akutz** [comment]
Hi @clintonskitson / @kacole2,

LMGTFY :)

(hint: it's the `-b` flag for sudo)

```
[0]akutz@pax:rexray$ ssh -t 192.168.120.16 rexray status
REX-Ray is stopped
Connection to 192.168.120.16 closed.

[0]akutz@pax:rexray$ ssh -t 192.168.120.16 sudo -b rexray start --force
Connection to 192.168.120.16 closed.

[0]akutz@pax:rexray$ ssh -t 192.168.120.16 rexray status
REX-Ray is running at pid 1258
Connection to 192.168.120.16 closed.

[0]akutz@pax:rexray$ ssh -t 192.168.120.16 sudo rexray stop
Shutting down REX-Ray...SUCCESS!
Connection to 192.168.120.16 closed.

[0]akutz@pax:rexray$ ssh -t 192.168.120.16 rexray status
REX-Ray is stopped
Connection to 192.168.120.16 closed.
```

**Comment by akutz** [comment]
Hi @clintonskitson,

By the way, I'm unable to reproduce the issue you're seeing with `tee`. Here is is with and without the `--force` flag:

```
[0]akutz@ubuntu:~$ sudo rexray start | sudo tee /tmp/temp
Oops, an error occured!

  no storage drivers initialized

To correct the error please review:

  - Debug output by using the flag -l debug
  - The REX-ray website at https://github.com/emccode/rexray
  - The on↓ine he↓p be↓ow

Start the service

Usage: 
  rexray service start [flags]

Flags:
      --client="": Socket the daemon uses to communicate to the client
      --force[=false]: Forces the service to start, ignoring errors
  -f, --foreground[=false]: Starts the service in the foreground

Global Flags:
  -c, --config="/root/.rexray/config.yaml": The REX-Ray configuration file
  -?, --help[=false]: Help for start
  -h, --host="tcp://:7979": The REX-Ray service address
  -l, --logLevel="info": The log level (panic, fatal, error, warn, info, debug)
  -v, --verbose[=false]: Print verbose help information

[0]akutz@ubuntu:~$ sudo rexray start --force | sudo tee /tmp/temp
Starting REX-Ray...SUCESS!

  The REX-Ray daemon is now running at PID 1515. To
  shutdown the daemon execute the following command:

    sudo /usr/bin/rexray stop

[0]akutz@ubuntu:~$ 
```

**Comment by clintonskitson** [comment]
`/docker-machine ssh testt5 "sudo rexray start > /dev/null 2>&1"`

**Comment by akutz** [comment]
Hi @clintonskitson,

Well, *that* isn't at all the example given with `tee`. `tee` was piping stdout and this is sending stderr to stdout and reassigning its file descriptor to /dev/null. Totally different. 

And yet, it still won't hang for me:

```
[1]akutz@pax:rexray$ ssh 192.168.120.16 "sudo rexray start --force > /dev/null 2>&1"

[1]akutz@pax:rexray$ ssh 192.168.120.16 sudo "rexray start --force > /dev/null 2>&1"
```

I mean, it fails, because the above command isn't correct. `sudo` does not have to be quoted as SSH doesn't require quoting, but the rest should if you want to redirect the descriptors. Interestingly if you remove the crazy descriptor stuff, you get the error you'll get with SSH's `-t` flag:

```
[1]akutz@pax:rexray$ ssh 192.168.120.16 sudo "rexray start --force"
sudo: sorry, you must have a tty to run sudo
```

Still, what I outlined above should probably work. You need to allocate a terminal first of all by using the `-t` flag for SSH and then use the `-b` flag for `sudo`:

```
/docker-machine ssh -t testt5 sudo -b rexray start
```

**Comment by akutz** [comment]
Hi @clintonskitson,

I will admit I'm not sure what dockermachine is doing, but anything it does to screw things up is on it. I've yet to reproduce any example of it hanging with a redirected file descriptor, so until I'm able to do that or someone can provide an example that is reproducible, I'm unable to address the issue.

**Comment by clintonskitson** [comment]
Closing the issue, the latest `unstable` release does not have this issue.
